File: extractor/extractor/strategies/norit_rechnung.py
from __future__ import annotations

import re

import fitz

from extractor.models import ExtractionResult, LineItem
from extractor.strategies.base import BaseStrategy
from extractor.utils import parse_de_number

_POS = re.compile(r"^\d{3}$")
_NET = re.compile(r"^(?P<net>[\d.,]+)\s+EUR\s*$", re.IGNORECASE)
_QTY = re.compile(
    r"^(?P<qty>[\d.,]+)\s+(?P<unit>m²|m2|St|kg|l|qm)\s*$", re.IGNORECASE
)
_UNIT_PRICE = re.compile(
    r"^(?P<price>[\d.,]+)\s+EUR\s*/\s*(?P<per>\S+)\s*$", re.IGNORECASE
)
_ARTNR = re.compile(r"^\d{8}$")
_SKIP = (
    "Pos",
    "Nettowert",
    "Einzelpreis",
    "Artikel",
    "Abw.",
    "Menge",
    "Übertrag",
    "Zolltarif",
    "Produkt",
    "CoC-",
    "Länge:",
    "Breite:",
    "Charge:",
    "VPE:",
    "Abmessung:",
    "Artikelnummer:",
    "EUR",
)


class NoritRechnungStrategy(BaseStrategy):
    layout_id = "norit_rechnung"

    def matches(self, doc: fitz.Document) -> bool:
        text = doc[0].get_text()
        return "Rechnungsnummer:" in text and "Einzelpreis" in text

    def extract(self, doc: fitz.Document, source_pdf: str) -> ExtractionResult:
        lines: list[str] = []
        for page in doc:
            lines.extend(line.strip() for line in page.get_text().splitlines())

        items: list[LineItem] = []
        i = 0
        while i < len(lines):
            if not _POS.match(lines[i]):
                i += 1
                continue
            if i + 1 >= len(lines):
                break
            net_match = _NET.match(lines[i + 1])
            if not net_match:
                i += 1
                continue

            position = lines[i]
            line_total = parse_de_number(net_match.group("net"))
            desc_parts: list[str] = []
            quantity = None
            unit = None
            unit_price = None
            article_number = None

            j = i + 2
            while j < len(lines):
                line = lines[j]
                if _POS.match(line) and j + 1 < len(lines) and _NET.match(lines[j + 1]):
                    break
                if line.startswith(_SKIP):
                    j += 1
                    continue
                qty_match = _QTY.match(line)
                if qty_match:
                    quantity = parse_de_number(qty_match.group("qty"))
                    unit = qty_match.group("unit")
                    j += 1
                    continue
                price_match = _UNIT_PRICE.match(line)
                if price_match:
                    unit_price = parse_de_number(price_match.group("price"))
                    j += 1
                    continue
                if _ARTNR.match(line):
                    article_number = line
                    j += 1
                    continue
                if line and not line.endswith("EUR"):
                    desc_parts.append(line)
                j += 1

            items.append(
                LineItem(
                    position=position,
                    article_number=article_number,
                    description=" ".join(desc_parts).strip(),
                    quantity=quantity,
                    unit=unit,
                    unit_price=unit_price,
                    line_total=line_total,
                )
            )
            i = j

        return ExtractionResult(
            layout_id=self.layout_id, source_pdf=source_pdf, items=items
        )

File: desktop/src/extractor/strategies/norit_rechnung.ts
import type { ExtractionResult, LineItem } from "../models";
import { parseDeNumber } from "../utils";
import type { Strategy } from "./base";

const POS = /^\d{3}$/;
const NET = /^(?<net>[\d.,]+)\s+EUR\s*$/i;
const QTY = /^(?<qty>[\d.,]+)\s+(?<unit>m²|m2|St|kg|l|qm)\s*$/i;
const UNIT_PRICE = /^(?<price>[\d.,]+)\s+EUR\s*\/\s*(?<per>\S+)\s*$/i;
const ARTNR = /^\d{8}$/;
const SKIP = [
  "Pos",
  "Nettowert",
  "Einzelpreis",
  "Artikel",
  "Abw.",
  "Menge",
  "Übertrag",
  "Zolltarif",
  "Produkt",
  "CoC-",
  "Länge:",
  "Breite:",
  "Charge:",
  "VPE:",
  "Abmessung:",
  "Artikelnummer:",
  "EUR",
] as const;

function startsWithAnySkipPrefix(line: string): boolean {
  return SKIP.some((p) => line.startsWith(p));
}

export function extractFromLines(lines: string[], source_pdf: string): ExtractionResult {
  const items: LineItem[] = [];

  let i = 0;
  while (i < lines.length) {
    const posLine = lines[i] ?? "";
    if (!POS.test(posLine)) {
      i += 1;
      continue;
    }

    if (i + 1 >= lines.length) break;
    const netMatch = NET.exec(lines[i + 1] ?? "");
    if (!netMatch?.groups) {
      i += 1;
      continue;
    }

    const position = posLine;
    const line_total = parseDeNumber(netMatch.groups.net ?? "");
    const descParts: string[] = [];
    let quantity: number | null = null;
    let unit: string | null = null;
    let unit_price: number | null = null;
    let article_number: string | null = null;

    let j = i + 2;
    while (j < lines.length) {
      const line = lines[j] ?? "";

      // next item header
      const nextPos = POS.test(line);
      const nextNet = j + 1 < lines.length && NET.test(lines[j + 1] ?? "");
      if (nextPos && nextNet) break;

      if (startsWithAnySkipPrefix(line)) {
        j += 1;
        continue;
      }

      const qtyMatch = QTY.exec(line);
      if (qtyMatch?.groups) {
        quantity = parseDeNumber(qtyMatch.groups.qty ?? "");
        unit = (qtyMatch.groups.unit ?? "").trim() || null;
        j += 1;
        continue;
      }

      const priceMatch = UNIT_PRICE.exec(line);
      if (priceMatch?.groups) {
        unit_price = parseDeNumber(priceMatch.groups.price ?? "");
        j += 1;
        continue;
      }

      if (ARTNR.test(line)) {
        article_number = line;
        j += 1;
        continue;
      }

      if (line && !line.endsWith("EUR")) {
        descParts.push(line);
      }
      j += 1;
    }

    items.push({
      position,
      article_number,
      description: descParts.join(" ").trim(),
      quantity,
      unit,
      unit_price,
      line_total,
    });

    i = j;
  }

  return { layout_id: "norit_rechnung", source_pdf, items };
}

export const NoritRechnungStrategy: Strategy = {
  layout_id: "norit_rechnung",
  matchesPage0Text: (page0Text: string) => page0Text.includes("Rechnungsnummer:") && page0Text.includes("Einzelpreis"),
  extract: (pdf: { pages: { lines: string[] }[] }, source_pdf: string) => {
    const lines: string[] = [];
    for (const page of pdf.pages) for (const line of page.lines) lines.push(line.trim());
    return extractFromLines(lines, source_pdf);
  },
};

